samza-core/src/main/java/org/apache/samza/coordinator/InputStreamRegexMonitor.java

```diff
@@ -0,0 +1,242 @@
+   * @param metadataCache     the metadata cache which will be used to fetch metadata for partition counts.
+   * @param metrics           the metrics registry to which the metrics should be added.
+   * @param inputRegexMonitorPeriodMs the period at which the monitor will check each input-regex
+   * @param monitorCallback   the callback method to be invoked when new input stream matching regex is detected
+   */
+  public InputStreamRegexMonitor(Set<SystemStream> streamsToMonitor, Map<String, Pattern> systemRegexesToMonitor,
+      StreamMetadataCache metadataCache, MetricsRegistry metrics, int inputRegexMonitorPeriodMs,
+      Callback monitorCallback) {
+    this.streamsToMonitor = streamsToMonitor;
+    this.systemRegexesToMonitor = systemRegexesToMonitor;
+    this.metadataCache = metadataCache;
+    this.callbackMethod = monitorCallback;
+    this.inputRegexMonitorPeriodMs = inputRegexMonitorPeriodMs;
+
+    // Pre-populate the gauges
+    Map<String, Gauge<Integer>> mutableGauges = new HashMap<>();
+    for (String systemToMonitor : systemRegexesToMonitor.keySet()) {
+      Gauge gauge = metrics.newGauge("job-coordinator", String.format("%s-new-input-streams", systemToMonitor), 0);
+      mutableGauges.put(systemToMonitor, gauge);
+    }
+    gauges = Collections.unmodifiableMap(mutableGauges);
+
+    log.info("Created {} with inputRegexMonitorPeriodMs: {} and systemRegexesToMonitor: {}", this.getClass().getName(),
+        this.inputRegexMonitorPeriodMs, this.systemRegexesToMonitor);
+  }
+
+  /**
+   * Starts the monitor.
+   */
+  public void start() {
+    synchronized (lock) {
+      switch (state) {
+        case INIT:
+          if (inputRegexMonitorPeriodMs > 0) {
+            schedulerService.scheduleAtFixedRate(new Runnable() {
+              @Override
+              public void run() {
+                monitorInputRegexes();
+              }
+            }, 0, inputRegexMonitorPeriodMs, TimeUnit.MILLISECONDS);
+          }
+          state = State.RUNNING;
+          break;
+
+        case RUNNING:
+          // start is idempotent
+          return;
+
+        case STOPPED:
+          throw new IllegalStateException("InputStreamRegexMonitor was stopped and cannot be restarted.");
+      }
+    }
+  }
+
+  /**
+   * Stops the monitor. Once it stops, it cannot be restarted.
+   */
+  public void stop() {
+    synchronized (lock) {
+      // We could also wait for full termination of the scheduler service, but it is overkill for
+      // our use case.
+      schedulerService.shutdownNow();
+
+      state = State.STOPPED;
+    }
+  }
+
+  private void monitorInputRegexes() {
+    log.debug("Running monitorInputRegexes");
+
+    try {
+      // obtain the list of SysStreams that match given patterns for all systems
+      Set<SystemStream> inputStreamsMatchingPattern = new HashSet<>();
+
+      // For each input system, for which we have a regex to monitor
+      for (String systemName : this.systemRegexesToMonitor.keySet()) {
```

Comment:
What do you think of initializing the set of streams to monitor when creating this class? Any deviation from this initial set, then results in the callback being invoked